src/Dialect/ONNX/ONNXOps.cpp

```diff
@@ -2285,112 +2285,190 @@ LogicalResult ONNXPadOp::inferShapes(
 // Unsqueeze
 //===----------------------------------------------------------------------===//
 
-LogicalResult ONNXUnsqueezeOp::inferShapes(
-    std::function<void(mlir::Region &)> doShapeInference) {
-  return emitError("Not implemented");
+// Update axes attribute so that it contains only positive values.
+// Helper functions for both Unsqueeze and Squeeze Ops
+template <typename Op>
+void updateNegativeAxis(Op *op, ArrayRef<int64_t> axes) {
```

Comment:
Do we really need to change the IR for axes if the shape info is stored in Index Expr?

Reply:
The original ShapeHelper functions for both Squeeze and Unsqueeze expect the axis values to be normalized during Shape inference such that there are no negative values. I lifted the block of code that updates the negative axis such that it's compatible with both ArrayAttr from V11 ops, and constant dense element attr from V13 ops.

Reply:
For constant, negative value can be handled in ShapeHelper. For dynamic value, the compare and select is anyway needed in ShapeHelper.

Reply:
I noticed this thread is marked as resolved, but do you want the handling of negative axes to be moved to ShapeHelper, or leave it as is at this stage? Thanks

Reply:
Just leave it as is.